libmscore/easeInOut.cpp

```diff
@@ -0,0 +1,116 @@
+// ease-out transfer curve for bends, glissandi and portamenti in MuseScore. The function computes
+// Y from X by performing a cubic root finding to compute t from X on the X component of the
+// transfer curve and then a cubic Bezier evaluation to compute Y from t on the Y component of the
+// transfer curve. The code could be simplified because the Bezier curve is constrained as a well
+// defined transfer function. It is not suitable for general or arbitrary Bezier curve work. -YP
+//-------------------------------------------------------------------------------------------------
+qreal EaseInOut::tFromX(const qreal x) const
+{
+    const qreal pi = 3.14159265358979323846;
+    qreal c = -x;
+    qreal w1 = _easeIn - x;
+    qreal w2 = 1.0 - _easeOut - x;
+    qreal d = x + 3.0 * w1 - 3.0 * w2 + (1.0 - x);
+    qreal a = (-3.0 * x - 6.0 * w1 + 3 * w2) / d;
+    qreal b = (3.0 * x + 3.0 * w1) / d;
+    c = -x / d;
+    qreal a2 = a * a;
+    qreal p = (3.0 * b - a2) / 3.0;
+    qreal q = (2.0 * a2 * a - 9.0 * a * b + 27.0 * c) / 27.0;
+    qreal discr = (q * q) / 4.0 + (p * p * p) / 27.0;
+    qreal t = 0.0;
+    if (discr < 0.0) {
+        // Case that happens about 75% of the time --> 3 roots.
+        qreal mp3 = -p / 3.0;
+        qreal r = std::sqrt(mp3 * mp3 * mp3);
+        qreal phi = std::acos(std::min(std::max(-1.0, -q / (2.0 * r)), 1.0));
+        qreal t1 = 2.0 * std::pow(r, 1.0 / 3.0);
+        // Find the only root that is within the 0 to 1 interval.
+        t = t1 * std::cos(phi / 3.0) - a / 3.0;
+        if (0.0 > t || t > 1.0) {
+            t = t1 * std::cos((phi + 2.0 * pi) / 3.0) - a / 3.0;
+            if (0.0 > t || t > 1.0) {
+                t = t1 * std::cos((phi + 4.0 * pi) / 3.0) - a / 3.0;
+            }
+        }
+    } else if (discr > 0.0) {
+        // Case that happens about 25% of the time --> 1 root.
+        qreal q2 = q / 2.0;
+        qreal sd = std::sqrt(discr);
+        qreal u = std::pow(-q2 + sd, 1.0 / 3.0);
+        qreal v = std::pow(q2 + sd, 1.0 / 3.0);
+        t = u - v - a / 3.0;
+    } else {
+        // Case that happens extremely rarely --> 2 roots.
+        qreal q2 = q / 2.0;
+        qreal u = q2 < 0.0 ? std::pow(-q2, 1.0 / 3.0) : -std::pow(q2, 1.0 / 3.0);
+        // Find the only root that is within the 0 to 1 interval.
+        t = 2.0 * u - a / 3.0;
+        if (0.0 > t || t > 1.0) {
+            t = -u - a / 3.0;
+        }
+    }
+    return t;
+}
+
+//-------------------------------------------------------------------------------------------------
+// This function is even more simplified and optimized because all the Bezier control points are
+// constant. Thus, when computing the Y root there is only one case and the math simplifies to the
+// following simple expression.
+//-------------------------------------------------------------------------------------------------
+qreal EaseInOut::tFromY(const qreal y) const
+{
+      if (y == 0.0 || y == 1.0)
```

Comment:
Same here, don't use floating <code>==</code> but use <code>qFuzzyCompare</code> instead.

Reply:
0.0 and 1.0 are represented precisely in float format. The trouble is when there is a fractional part on the number. In this case, anyway, those are precise end conditions. Any other y values will require a Bezier interpolation.

Reply:
Maybe, but equality is not really. Is 0.0000000001 == 0.0?

Reply:
I don't understand. 0.00000000000001 is not equal to 0.0. Do you suggest that I truncate small values to zero?

Reply:
yes

Reply:
That is what `qFuzzyCompare()`does

Reply:
When using float (or double) calculation, the result is almost always rounded because the the actual result doesn't "fit" in the float/double format. So, rounding errors of +/- 1 bit are most likely.
<code>qFuzzCompare</code> indeed tries to take this "one bit errors" into account. See [here](https://floating-point-gui.de/errors/comparison/) for a nice explanation.

Reply:
BTW, my preferred reference concerning floating point error management is from the book "[Physically Based Rendering: From Theory to Practice, Chapter 3.9 - Managing Rounding Errors](http://www.pbr-book.org/3ed-2018/Shapes/Managing_Rounding_Error.html)". I invite you to read this chapter for an in-depth explanation.

Reply:
The first link is about floating comparisons, while your link is about rounding errors. So, different subjects
 Nevertheless, thanks, very interesting. I will certainly go over it. Reminds me to my old study years, back in the seventies (but without a nice book!) :smile: .

Reply:
@njvdberg To me, this is the same subject. Rather than focusing on the application of a float comparison rule, I focus on the consequences of using a float comparison vs letting the error go through. This is something I learned in 3D rendering where performance is of paramount importance and the heavy arsenal of error management (of which qFuzzyCompare is a member) is avoided by making the code robust to errors. Error management code is used only in the very few places where it is absolutely required.

So I spent one month of leisure time analyzing the boundary conditions of the Bezier functions using Mathematica before I started looking at MuseScore code. I tried to argue, here, that there are no consequences in breaking this rule in this function because even if the zero discriminant is only very close to zero due to float errors, the path that will be followed will still work correctly. This is even more so given the constraints in which the cubic Bezier is used as transfer function.

What turned me off, however, is the requirement to use an int in a for loop. I fully accept the example where the for loop will loop through one too few. That is a good example and a very possible outcome. A blatant example of why code review is good. However, IMO, the right solution is to add one half of the increment step to the stop condition. Not to use int. Using int in interpolation functions results in, for example, the MuseScore bend function not reaching its true final note most of the time.

I understand you are a 50 years experienced Fortran and C++ (and probably more) programmer. I respect you for this. I'm also an experienced programmer. 40 years with Assembler, Pascal, Lisp, Mathematica, C/C++, Javascript (a little bit of Fortran as student and a few other arcane languages). My specialty is in 3D computer graphics and rendering, of which the last 10 years in physically-based rendering.

Reply:
@ypoissant I started with Fortran (and Algol) during my studies and used Fortran in my first job for scientific programs. Last 20 years I worked as a CAD engineer (semiconductors) and used mainly SKILL/SKILL++ (which is a Lisp dialect). There I  spent days on programs showing a very strange result (or even didn't come to an end) and often it was because in the algorithm there was a floating (not) equal test. E.g. in one case I found the string "1n" and "1e-9" (which is the same) give a different result. it turns out "1n" was internally replaced by an expression 1.0*1e-9 which which changes in the last bit from 1e-9. This kind of issues I've seen several times over the years. And sometimes in code which worked fine for years. That's the reason I'm always triggered when I see this kind of test. No intent to offend you, and if I did, my apologies.

My worry is still, in this particular case, depending on the calculation, <code>y</code> is 0.0 but due to the limited number of bits for the mantissa the last bit flips so the result is not exactly 0.0 but something like 1.11022e-16 (on my computer). then the test <code>y == 0.0</code> will fail but <code>qFuzzyCompare(y, 0.0)</code> not since this functions tries to take this last mantissa bit into account.